Adds Japanese extraction function and post-processing.

lfashby opened the pull request:
This is my attempt to write an extraction function for Japanese. It isn't very pretty, but it seems to work. I've included the Japanese data that I scraped using this extraction function and a post-processing step for splitting the Japanese TSVs into hiragana and katakana. 

The names of the TSVs are `jpn_hira_...` and `jpn_kana_...` because "Hira" and "Kana" are the [ISO 15294 ](https://en.wikipedia.org/wiki/ISO_15924) designation for the scripts but I can change them if we'd like.

I can also add additional comments in `extract/jpn.py` if they would be helpful. 

I'm sure there will be suggestions to change certain things in `jpn.py`. I won't have too much free time until Wednesday night so I'll address the changes then. 

Resolves #73

kylebgorman commented:
Generally endorsing this though I would like Jackson to have the final word. 

I don't really see anything here such that it's obvious that it can be done better. And 15k entries is nothing to sneer at.

jacksonllee commented on wikipron/extract/jpn.py:
```diff
@@ -0,0 +1,229 @@
+def _check_etymologies(request: requests.Response):
+    count = 0
+    for a_element in request.html.xpath(_TOC_ETYMOLOGY_XPATH_SELECTOR):
+        count += 1
```
Would `count = len(request.html.xpath(_TOC_ETYMOLOGY_XPATH_SELECTOR))` simply work? I can't remember if `request.html.xpath` already returns a materialized array of things that would allow `len`. If not, and if we're keeping this `for` loop, `a_element` could simply be `_` as it's not used.

kylebgorman replied:
If it's not an actual `len`-able object the hack is:

    count = sum(1 for _ in request.html.xpath(...))

jacksonllee commented on wikipron/extract/jpn.py:
```diff
@@ -0,0 +1,229 @@
+        prons = []
+        for pron in yield_pron(upper_pron_ele, IPA_XPATH_SELECTOR, config):
+            prons.append(pron)
+        yield prons
```
Would the following work?

```python
prons = list(yield_pron(upper_pron_ele, IPA_XPATH_SELECTOR, config))
yield prons
```

kylebgorman replied:
can't you do `yield from yield_pron(...)`?

jacksonllee replied:
IIUC, the code was trying to yield a list of prons, rather than yield the prons one by one, and hence my suggested refactoring. That being said, now that I'm reading this again in the larger context, I think Lucas was trying to generate both "lower" and "upper" prons, chain them together, and then yield the prons one at a time. In this case, I think there's a more natural way to refactor the code. The first part is what Kyle has suggested: We should do `yield from yield_pron(...)` here, so that `_yield_jpn_upper_pron` yields the prons one by one. Other parts are in my updated comments below.

lfashby replied:
The idea in both of these pron yielding functions is that I want to exhaustively run `yield_pron()` on each iteration of the upper for loops in each of the pron yielding functions. 

When the next `next(prons)` in `_combine_word_pron_pairs()` gets called, the upper for loop in both pron yielding functions gets ‘incremented’ (if that’s the right word?) and we target the subsequent word-pron pair group on the page and grab the prons in that group.

I am trying to potentially yield a list of prons in each function. Jackson's original `list()` suggestion is a simplified way of exhaustively running `yield_pron()`, but it still leaves this:
```
        try:
            prons += next(upper_prons)
        except StopIteration:
            pass
```
block in `_yield_jpn_lower_pron`. (If this comment and the comment I left on the `itertools.chain()` thread don't make any sense then let me know.)

jacksonllee commented on wikipron/extract/jpn.py:
```diff
@@ -0,0 +1,229 @@
+    for pron_element in request.html.xpath(pron_path):
+        prons = []
+        for pron in yield_pron(pron_element, IPA_XPATH_SELECTOR, config):
+            prons.append(pron)
```
`prons = list(yield_pron(pron_element, IPA_XPATH_SELECTOR, config))` ?

jacksonllee commented:
Forgot to mention this: Should we remove `languages/jpn_wik/` and everything in this subdirectory?

kylebgorman commented:
Yes, IMO, that's now handled nicely.

On Tue, Nov 26, 2019 at 5:28 PM Jackson L. Lee <notifications@github.com>
wrote:

> Forgot to mention this: Should we remove languages/jpn_wik/ and
> everything in this subdirectory?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kylebgorman/wikipron/pull/109?email_source=notifications&email_token=AABG4ONT5LGUKI2772GWMIDQVWPHPA5CNFSM4JRSFN62YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFHUYLY#issuecomment-558844975>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABG4OP6KQYLENG3GIYGS4TQVWPHPANCNFSM4JRSFN6Q>
> .
>

kylebgorman commented on languages/wikipron/README.md:
```diff
@@ -91,8 +91,10 @@
-| [TSV](tsv/jpn_phonetic.tsv) | jpn | Japanese | Japanese | False | Phonetic | 24746 |
+| [TSV](tsv/jpn_hira_phonemic.tsv) | jpn | Japanese | Japanese (Hiragana) | False | Phonemic | 137 |
+| [TSV](tsv/jpn_hira_phonetic.tsv) | jpn | Japanese | Japanese (Hiragana) | False | Phonetic | 14494 |
+| [TSV](tsv/jpn_kana_phonemic.tsv) | jpn | Japanese | Japanese (Katakana) | False | Phonemic | 1 |
```
+1 we normally do this and it's mentioned in the README change

lfashby replied:
My bad, I changed `split_jpn` to automatically remove the small files (and should probably add that behavior to `split_hbs` just to be consistent), but forgot to re-run `generate_summary.py`.

kylebgorman commented on wikipron/extract/jpn.py:
```diff
@@ -0,0 +1,229 @@
+            associated_prons = next(prons)
+        except StopIteration:
+            pass
+        yield from (zip(itertools.repeat(word), associated_prons))
```
+1

jacksonllee replied:
If `prons` returned by `_yield_jpn_lower_pron` yields prons one by one, then we don't need to catch `StopIteration` here:

```python
for word in words:
    yield from zip(itertools.repeat(word), prons)
```

kylebgorman commented:
> Forgot to mention this: Should we remove `languages/jpn_wik/` and everything in this subdirectory?

Yes please, though it doesn't have to be done in this PR if you don't feel like it.

jacksonllee commented on wikipron/extract/jpn.py:
```diff
@@ -0,0 +1,229 @@
+        # entries on a page at the same time. Doing so would make it difficult
+        # to connect words to their prons appropriately. We only want to grab
+        # the prons that are linked to the word we have scraped.
+        yield prons
```
We probably don't have to catch `StopIteration`, if we use `itertools.chain`:

```python
for pron_element in request.html.xpath(pron_path):
    lower_prons = yield_pron(...)
    upper_prons = _yield_jpn_upper_pron(...)
    # The following won't throw an error even if either or both of lower_prons and upper_prons are empty
    prons = itertools.chain(lower_prons, upper_prons)
    yield from prons
```

lfashby replied:
I like the `itertools.chain()` approach, but I'm hesitant to exhaustively run `_yield_jpn_upper_pron(...)` each iteration of the loop. Then it would collect all the upper prons from all the Etymologies each time we grab the lower pron or prons for a specific word, potentially leading to a mismatch of words and pronunciations. (I say potentially because I've never seen any page with multiple etymologies that have multiple <ul>'s.)

lfashby commented:
I addressed most of the suggestions and added some comments to `jpn.py`. If there are any ideas on how to refactor the pron yielding functions in light of my responses to your comments then let me know. I'll be out for the rest of the day but will be around tomorrow.

lfashby commented:
Great. I learned a lot while working on this so I think it would be rewarding to try and rework it over winter break when I have some free time.
